Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferProxy.cpp

```diff
@@ -41,32 +41,78 @@
 namespace WebKit {
 using namespace WebCore;
 
+
+class RemoteImageBufferProxyFlushFence : public ThreadSafeRefCounted<RemoteImageBufferProxyFlushFence> {
+    WTF_MAKE_NONCOPYABLE(RemoteImageBufferProxyFlushFence);
+    WTF_MAKE_FAST_ALLOCATED;
+public:
+    static Ref<RemoteImageBufferProxyFlushFence> create(IPC::Semaphore semaphore)
+    {
+        return adoptRef(*new RemoteImageBufferProxyFlushFence { WTFMove(semaphore) });
+    }
+
+    ~RemoteImageBufferProxyFlushFence()
+    {
+        if (!m_signaled)
+            tracePoint(FlushRemoteImageBufferEnd, reinterpret_cast<uintptr_t>(this), 2u);
+    }
+
+    bool waitFor(Seconds timeout)
+    {
+        Locker locker { m_lock };
+        if (m_signaled)
+            return true;
+        m_signaled = m_semaphore.waitFor(timeout);
+        if (m_signaled)
+            tracePoint(FlushRemoteImageBufferEnd, reinterpret_cast<uintptr_t>(this), 0u);
+        return m_signaled;
+    }
+
+    void forceSignal()
+    {
+        Locker locker { m_lock };
+        if (m_signaled)
+            return;
+        m_signaled = true;
+        tracePoint(FlushRemoteImageBufferEnd, reinterpret_cast<uintptr_t>(this), 1u);
+        m_semaphore.signal();
+    }
+
+private:
+    RemoteImageBufferProxyFlushFence(IPC::Semaphore semaphore)
+        : m_semaphore(WTFMove(semaphore))
+    {
+        tracePoint(FlushRemoteImageBufferStart, reinterpret_cast<uintptr_t>(this));
+    }
+    Lock m_lock;
+    Condition m_condition;
+    bool m_signaled WTF_GUARDED_BY_LOCK(m_lock) { false };
+    IPC::Semaphore m_semaphore;
+};
+
 namespace {
 
 class RemoteImageBufferProxyFlusher final : public ThreadSafeImageBufferFlusher {
     WTF_MAKE_FAST_ALLOCATED;
 public:
-    RemoteImageBufferProxyFlusher(RemoteImageBufferProxyFlushState& flushState, DisplayListRecorderFlushIdentifier identifier)
-        : m_flushState(flushState)
-        , m_targetFlushIdentifier(identifier)
+    RemoteImageBufferProxyFlusher(Ref<RemoteImageBufferProxyFlushFence> flushState)
+        : m_flushState(WTFMove(flushState))
     {
     }
 
     void flush() final
     {
-        m_flushState->waitForDidFlush(m_targetFlushIdentifier, Seconds::infinity());
+        m_flushState->waitFor(RemoteDisplayListRecorderProxy::defaultSendTimeout);
```

Comment:
Previously we never timed out for flushing on a background thread (for rendering updates), and now we're adding a 3second timeout there.

I'm not super sure why we have different behaviour between threads currently, but it seems like changes to it should be done explicitly.

Assuming you think this is the right behaviour change, can you please add it to the commit message (or possibly even just land a separate one line commit, so this refactor becomes a no-op in terms of behaviour).

Reply:
#13873